checkbox-ng/plainbox/impl/new_resource.py

```diff
@@ -0,0 +1,711 @@
+    of the object. All attributes must refer to the same namespace.
+    """
+
+    CALLS_MEANING = {
+        "int": int,
+        "bool": bool,
+        "float": float,
+        "str": str,
+    }
+
+    def _get_namespace_args(self, args):
+        namespaces = {
+            arg.namespace for arg in args if hasattr(arg, "namespace")
+        }
+        if len(namespaces) > 1:
+            raise ValueError(
+                "Function call can access at most one namespace but this is "
+                "using: " + " ".join(namespaces)
+            )
+        elif len(namespaces) == 0:
+            raise ValueError(
+                "Function calls with no namespace are unsupported"
+            )
+        return namespaces.pop()
+
+    def __init__(self, parsed_ast):
+        try:
+            self.function_name = parsed_ast.func.id
+            self.function = self.CALLS_MEANING[parsed_ast.func.id]
+        except KeyError:
+            raise ValueError(
+                "Unsupported function {}".format(parsed_ast.func.id)
+            )
+
+        self.args = [getter_from_ast(arg) for arg in parsed_ast.args]
+        self.namespace = self._get_namespace_args(self.args)
+
+    def __call__(self, variable_object):
+        return self.function(*(arg(variable_object) for arg in self.args))
+
+    def __str__(self):
+        args = ",".join(str(x) for x in self.args)
+        return "{}({})".format(self.function_name, args)
+
+
+class AttributeGetter(NamespacedGetter):
+    def __init__(self, parsed_ast):
+        self.namespace = parsed_ast.value.id
+        self.variable = parsed_ast.attr
+
+    def __call__(self, variable_object):
+        # resources are free form, support variable names not being unifrom
+        try:
+            return variable_object[self.variable]
+        except KeyError:
+            return None
+
+    def __str__(self):
+        return "{}.{}".format(self.namespace, self.variable)
+
+
+class ConstantGetter(ValueGetter):
+    def __init__(self, parsed_ast):
+        self.value = parsed_ast.value
+
+    def __call__(self, *args):
+        return self.value
+
+    @classmethod
+    def from_unary_op(cls, parsed_ast):
+        if not isinstance(parsed_ast.op, ast.USub):
+            raise ValueError("Unsupported operator {}".format(parsed_ast))
+        operand = getter_from_ast(parsed_ast.operand)
+        if not isinstance(operand, ConstantGetter):
+            raise ValueError(
+                "`-` operator can't be applied to non-constant operands"
+            )
+        operand.value *= -1
+        return cls(operand)
+
+    def __str__(self):
+        return str(self.value)
+
+
+class NamedConstant(ConstantGetter):
+    constants = {
+        "DESKTOP_PC_PRODUCT": [
+            "Desktop",
+            "Low Profile Desktop",
+            "Tower",
+            "Mini Tower",
+            "Space-saving",
+            "All In One",
+            "All-In-One",
+            "AIO",
+        ]
+    }
```

Comment:
This was more of a conversation starter, we can have a constants.json file, I wanted that, but someone else wanted constants to be namespaced. I dont think that is a good idea so that is why the constant thing is hardcoded

Reply:
offloaded the constants to a file